File: scripts/spike/looksright/judge.py
#!/usr/bin/env python3
"""Throwaway (task: art-judge). Reads the frames judge.mjs shot and reports contrast
AT THE REAL STANDS, rim thickness in css px, and floor obscured by the boss.

WCAG method, same as docs/art/legibility.py: sRGB -> linear, Y = .2126R+.7152G+.0722B,
K = (hi+.05)/(lo+.05).
"""
import json
import numpy as np
from PIL import Image

OUT = '/home/anshtyagi/Documents/pixel-artgame/docs/art/shipped/judge'
SKINS = ['Cobalt', 'Nocturne', 'Argent']
SPRITE_W, SPRITE_H = 33, 42


def lin(c):
    c = np.asarray(c, float) / 255.0
    return np.where(c <= 0.04045, c / 12.92, ((c + 0.055) / 1.055) ** 2.4)


def Y(rgb):
    l = lin(rgb)
    return 0.2126 * l[..., 0] + 0.7152 * l[..., 1] + 0.0722 * l[..., 2]


def K(a, b):
    hi, lo = max(a, b), min(a, b)
    return (hi + 0.05) / (lo + 0.05)


img = lambda n: np.asarray(Image.open(f'{OUT}/{n}.png').convert('RGB'), float)
META = {m['name']: m for m in json.load(open(f'{OUT}/meta.json'))}

out = []
P = out.append


def px(m, x, y):
    c = m['geom']['ctm']
    return c['a'] * x + c['e'], c['a'] * y + c['f']


def seat_stats(name, plate):
    """Body mask = |frame - plate| inside each seat's sprite box; floor = the plate there."""
    m = META[name]
    f, p = img(name), img(plate)
    s = m['geom']['ctm']['a']
    d = np.abs(f - p).max(axis=2)
    rows = []
    for i, (wx, wy, skin, _) in enumerate(m['at']):
        if i == 0:
            continue  # parked local seat, wears the chevron
        cx, cy = px(m, wx, wy)
        # the sprite is drawn centred on the seat point; take a generous box and let the
        # diff mask pick out the body.
        x0, x1 = int(cx - SPRITE_W * s / 2) - 4, int(cx + SPRITE_W * s / 2) + 5
        y0, y1 = int(cy - SPRITE_H * s / 2) - 4, int(cy + SPRITE_H * s / 2) + 5
        if x0 < 0 or y0 < 0 or x1 > f.shape[1] or y1 > f.shape[0]:
            continue
        bf, bp = f[y0:y1, x0:x1], p[y0:y1, x0:x1]
        mask = d[y0:y1, x0:x1] > 12
        if mask.sum() < 120:
            continue
        body = float(np.median(Y(bf)[mask]))
        # floor = the plate under the box, excluding nothing: this is what the eye compares to
        floor = float(np.median(Y(bp)))
        # boundary ring: the mask's outer 2-px shell in the FRAME (the rim, if there is one)
        rows.append((body, floor, K(body, floor)))
    return rows


def report(sets):
    for setname, label in sets:
        P(f'\n-- {label} --')
        for skin in range(3):
            for cls, suffix in (('knight', ''), ('archer', '-arch')):
                nm = f'{setname}-s{skin}{suffix}'
                if nm not in META:
                    continue
                rows = seat_stats(nm, f'{setname}-plate')
                if not rows:
                    P(f'   {SKINS[skin]:9s} {cls:6s}  NO SEATS RESOLVED')
                    continue
                ks = sorted(r[2] for r in rows)
                bodies = [r[0] for r in rows]
                floors = [r[1] for r in rows]
                P(f'   {SKINS[skin]:9s} {cls:6s} n={len(rows):2d}  body Y {np.median(bodies):.4f}'
                  f'  floor Y {np.median(floors):.4f}  contrast median {np.median(ks):.2f}:1'
                  f'  worst {ks[0]:.2f}:1  best {ks[-1]:.2f}:1')


P('== contrast at the stands the chain actually produces (1920x1080) ==')
P('   body = median Y of the changed pixels in the seat box; floor = median Y of the same')
P('   box on the empty plate. n is the number of seats that resolved.')
report([('spawn', 'room A: the lobby spawn row, map.ts LOBBY_SPAWN_Y 832, x 208..664, 19 seats'),
        ('walk', 'room A: the aisle walk from the spawn row onto the gate block, 12 seats'),
        ('pools', 'room A: standing in the CENTRE of each of the ten torch pools, 10 seats'),
        ('pit', 'room B: the arena pit, y 440..596 across x 140..860, 19 seats')])

# ------------------------------------------------------------------ rim thickness
P('\n== rim light: measured off an 8x-zoomed knight, not asserted ==')
m = META['rim-s0']
s = m['geom']['ctm']['a']
f, p = img('rim-s0'), img('rim-plate')
d = np.abs(f - p).max(axis=2)
mask = d > 10
ys, xs = np.nonzero(mask)
if xs.size == 0:
    P('   SKIPPED: the 8x-zoom seat is off the page under the new camera; not re-measured.')
    print('\n'.join(out))
    open(f'{OUT}/../judge-report.txt', 'w').write('\n'.join(out) + '\n')
    raise SystemExit(0)
P(f'   1 world unit = {s:.3f} css px at this zoom (8x of the shipped 1.574)')
P(f'   changed-pixel bbox {xs.max()-xs.min()+1} x {ys.max()-ys.min()+1} px'
  f'  = {(xs.max()-xs.min()+1)/s:.1f} x {(ys.max()-ys.min()+1)/s:.1f} world units'
  f'  (sprite is {SPRITE_W}x{SPRITE_H})')
# The rim is the brightest ring just inside the mask boundary. Walk the mask's widest row
# and count how many px at each end are rim-coloured (much brighter than the floor).
mid = int(np.median(ys))
row = mask[mid]
xi = np.nonzero(row)[0]
lumrow = Y(f[mid])
floor = float(np.median(Y(p[mid][xi.min():xi.max()])))
runL = 0
for x in range(xi.min(), xi.max()):
    if lumrow[x] > floor * 3:
        runL += 1
    elif runL:
        break
P(f'   widest scanline y={mid}: floor Y {floor:.4f}; leading bright run {runL} px'
  f' = {runL/s:.2f} world units = {runL/s*1.574:.2f} css px at the shipped 1920 zoom')

# ------------------------------------------------------------------ boss over the pit
P('\n== room B: how much of the walkable pit the creature covers ==')
SH = '/home/anshtyagi/Documents/pixel-artgame/docs/art/shipped'
gm = {g['name']: g for g in json.load(open(f'{SH}/geometry.json'))}
c = gm['arena-1920']['geom']['ctm']
full = np.asarray(Image.open(f'{SH}/plate-arena-nohud-empty.png').convert('RGB'), float)
noboss = np.asarray(Image.open(f'{SH}/plate-arena-noboss.png').convert('RGB'), float)
bossmask = np.abs(full - noboss).max(axis=2) > 10
PIT_TOP, PIT_BOT = 384, 607
x0 = int(c['a'] * 0 + c['e'])
x1 = int(c['a'] * 1024 + c['e'])
y0 = int(c['a'] * PIT_TOP + c['f'])
y1 = int(c['a'] * (PIT_BOT + 1) + c['f'])
y0c, y1c = max(0, y0), min(full.shape[0], y1)
pit = bossmask[y0c:y1c, max(0, x0):min(full.shape[1], x1)]
P(f'   pit band on screen: x {x0}..{x1}, y {y0}..{y1} (clipped to {y0c}..{y1c})')
P(f'   creature covers {pit.mean():.1%} of the walkable pit rectangle'
  f'  ({pit.sum():,} of {pit.size:,} px)')
# the reference's creature over its own floor, for the same ratio
open(f'{OUT}/../judge-report.txt', 'w').write('\n'.join(out) + '\n')
print('\n'.join(out))

File: scripts/spike/looksright/judge.mjs
// Throwaway (task: art-judge). Places knights WHERE THE CHAIN ACTUALLY PUTS THEM --
// `map.ts` LOBBY_SPAWN row, the aisle walk to the gate, the ten torch-pool centres, and
// the arena pit -- one skin per frame, plus a matching empty plate, so body-vs-floor
// contrast is a pixel diff at the real stands instead of over a synthetic grid.
import { createRequire } from 'node:module';
const require_ = createRequire(process.env.PW_HOME + '/x.cjs');
const { chromium } = require_('playwright');
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { QUIET_ARGS } from '../launch.mjs';

const DIR = '/tmp/looksright-dist';
const OUT = '/home/anshtyagi/Documents/pixel-artgame/docs/art/shipped/judge';
const MIME = { '.html': 'text/html', '.js': 'text/javascript', '.css': 'text/css', '.svg': 'image/svg+xml' };
const srv = http.createServer((q, s) => {
  const f = path.join(DIR, q.url === '/' ? 'index.html' : q.url.split('?')[0]);
  if (!f.startsWith(DIR) || !fs.existsSync(f)) { s.writeHead(404); return s.end(); }
  s.writeHead(200, { 'content-type': MIME[path.extname(f)] || 'text/plain' });
  fs.createReadStream(f).pipe(s);
});
const PORT = 8931;
await new Promise((r) => srv.listen(PORT, r));
fs.mkdirSync(OUT, { recursive: true });

// --- the real stands, straight out of packages/client/src/map.ts -----------------
const SPAWN_MIN_X = 208, SPAWN_MAX_X = 664, SPAWN_Y = 832;
const GATE_CX = 512;
// seat 0 is the local player and wears the chevron, so it is parked off-set and never
// measured; every measured seat is index >= 1.
const PARK = [1000, 1000, 0, 0];

const spawnRow = [PARK];
for (let i = 0; i < 19; i++) {
  spawnRow.push([Math.round(SPAWN_MIN_X + (i * (SPAWN_MAX_X - SPAWN_MIN_X)) / 18), SPAWN_Y, 0, 0]);
}
// the walk: straight up the aisle from the spawn row onto the gate block (y 608..639)
const walk = [PARK];
for (let i = 0; i < 12; i++) walk.push([GATE_CX + (i % 2 ? 26 : -26), 832 - i * 18, 0, 0]);
// A grid across the whole LOBBY floor (map.ts LOBBY_TOP 688 .. LOBBY_BOT 1007, cols 2..61
// => world x 32..991). Replaces the old typed torch-pool list, which was coordinates from
// a WaitingRoom that no longer types any.
const pools = [PARK];
for (let r = 0; r < 4; r++) for (let c = 0; c < 5; c++) {
  if (pools.length >= 20) break;
  pools.push([90 + c * 210, 730 + r * 80, 0, 0]);
}
// arena pit: ZONE_ARENA y is PIT_TOP..PIT_BOT (384..655); entrances are y 512.
const pit = [PARK];
for (let r = 0; r < 4; r++) for (let c = 0; c < 5; c++) { if (pit.length >= 20) break; pit.push([140 + c * 180, 400 + r * 62, 0, 0]); }

const SETS = { spawn: spawnRow, walk, pools, pit };
const HIDE = '.hud,.dev,.hr-hud,[class*="hud"],[class*="dev"],.errorbar{display:none!important}';

const ONLY = (process.env.ONLY || '').split(',').filter(Boolean);
const CASES = [];
for (const [set, at] of Object.entries(SETS)) {
  const room = set === 'pit' ? 'arena' : 'lobby';
  for (const skin of [0, 1, 2]) {
    CASES.push({ name: `${set}-s${skin}`, room, at: at.map((p) => [p[0], p[1], skin, 0]) });
    CASES.push({ name: `${set}-s${skin}-arch`, room, at: at.map((p) => [p[0], p[1], skin, 1]) });
  }
  CASES.push({ name: `${set}-plate`, room, at: [PARK] });
}
// One big knight for a direct rim measurement: single seat, page zoomed 8x by CSS.
CASES.push({ name: 'rim-s0', room: 'lobby', at: [PARK, [512, 760, 0, 0]], zoom: 8 });
CASES.push({ name: 'rim-plate', room: 'lobby', at: [PARK], zoom: 8 });

const meta = [];
for (const c of (ONLY.length ? CASES.filter((x) => ONLY.some((o) => x.name.startsWith(o))) : CASES)) {
  const b = await chromium.launch({ channel: 'chrome', headless: false, args: [...QUIET_ARGS] });
  try {
    const pg = await b.newPage({ viewport: { width: 1920, height: 1080 }, deviceScaleFactor: 1 });
    await pg.goto(`http://127.0.0.1:${PORT}/`, { waitUntil: 'load' });
    await pg.waitForFunction('window.__ready === true', null, { timeout: 20000 });
    await pg.evaluate(async () => { await window.__store.signIn(); await window.__store.join(); });
    const opts = { at: c.at, bullets: 0 };
    await pg.evaluate(([room, o]) => window.__scene(room, o), [c.room, opts]);
    await pg.waitForTimeout(2200);
    await pg.evaluate(([room, o]) => window.__scene(room, o), [c.room, opts]);
    await pg.waitForTimeout(1400);
    await pg.addStyleTag({ content: HIDE });
    if (c.zoom) {
      await pg.evaluate((z) => {
        const svg = document.getElementById('stage').querySelector('svg');
        svg.style.transformOrigin = '0 0';
        svg.style.transform = `scale(${z})`;
      }, c.zoom);
    }
    await pg.waitForTimeout(300);
    const geom = await pg.evaluate(() => {
      const stage = document.getElementById('stage');
      const svg = stage?.querySelector('svg');
      const m = svg?.querySelector('#camera')?.getScreenCTM?.();
      const sr = stage?.getBoundingClientRect();
      return { stage: sr && { x: +sr.x.toFixed(1), y: +sr.y.toFixed(1), w: +sr.width.toFixed(1), h: +sr.height.toFixed(1) },
               ctm: m && { a: +m.a.toFixed(5), d: +m.d.toFixed(5), e: +m.e.toFixed(2), f: +m.f.toFixed(2) } };
    });
    await pg.screenshot({ path: path.join(OUT, c.name + '.png'), timeout: 15000 });
    meta.push({ name: c.name, room: c.room, at: c.at, zoom: c.zoom || 1, geom });
    console.log(c.name, JSON.stringify(geom.ctm));
  } catch (e) { console.log(c.name, 'FAILED', String(e).slice(0, 200)); }
  finally { await b.close(); }
}
{ const f = path.join(OUT, 'meta.json'); const old = fs.existsSync(f) ? JSON.parse(fs.readFileSync(f)) : [];
  const byName = new Map(old.map((m) => [m.name, m])); meta.forEach((m) => byName.set(m.name, m));
  fs.writeFileSync(f, JSON.stringify([...byName.values()], null, 2)); }
srv.close();
